jobRun.go

```diff
@@ -0,0 +1,284 @@
+			return err
+		}
+		return nextJobRun.insertGet(tx)
+	})
+	return nextJobRun, txErr
+}
+
+// retryOnTimeout does that
+func (j *JobRun) retryOnTimeout(db *gorm.DB, instanceID int64) (*JobRun, error) {
+
+	// nothing should happend if the job run has completed or has closed
+	if j.ClosedAt.Valid || j.ClosedBy.Valid {
+		return nil, nil
+	}
+	if j.RunCompletedAt.Valid || j.RetriesOnTimeoutCount >= j.RetriesOnTimeoutLimit {
+		return nil, nil
+	}
+
+	// retry the job run if it produced an error
+	nextJobRun := j.CloneReset(instanceID)
+	nextJobRun.RetriesOnTimeoutCount++
+
+	txErr := db.Transaction(func(tx *gorm.DB) error {
+		if err := j.close(tx, instanceID); err != nil {
+			return err
+		}
+		return nextJobRun.insertGet(tx)
+	})
+	return nextJobRun, txErr
+}
+
+// reschedule the job run
+func (j *JobRun) reschedule(db *gorm.DB, instanceID int64, schedules map[string]ScheduleFunc) (*JobRun, error) {
+
+	// nothing should happend if the job run does not need scheduling or has closed
+	if j.ClosedAt.Valid || j.ClosedBy.Valid {
+		return nil, nil
+	}
+	if !j.needsScheduling() {
+		return nil, nil
+	}
+
+	schedule, exists := schedules[j.Schedule.String]
+	if !exists {
+		return nil, errors.New("cannot reschedule job run, schedule does not exist")
+	}
+
+	nextJobRun := j.CloneReset(instanceID)
+	nextJobRun.RunAt = schedule(j.RunCompletedAt.Time)
+	nextJobRun.RetriesOnErrorCount = 0
+	nextJobRun.RetriesOnTimeoutCount = 0
+
+	txErr := db.Transaction(func(tx *gorm.DB) error {
+		if err := j.close(tx, instanceID); err != nil {
+			return err
+		}
+		return nextJobRun.insertGet(tx)
+	})
+	return nextJobRun, txErr
+}
+
+func (j *JobRun) needsScheduling() bool {
+	return j.Schedule.Valid && (!j.RunLimit.Valid || j.RunLimit.Int64 > int64(j.RunCount))
+}
+
+func (j *JobRun) check(
+	jobs map[string]*JobContainer,
+	schedules map[string]ScheduleFunc,
+) error {
+
+	jobC, exists := jobs[j.Job]
+	if !exists {
+		return errors.New("cannot run job. job '" + j.Job + "' does not exist")
+	}
+	if err := jobC.jobFunc.check(j.JobArgs); err != nil {
+		return err
+	}
+
+	if j.needsScheduling() {
+		_, exists := schedules[j.Schedule.String]
+		if !exists {
+			return errors.New("cannot schedule job. schedule '" + j.Schedule.String + "' missing")
+		}
+	}
+
+	return nil
+}
+
+func (j *JobRun) logger(logger logc.Logger) logc.Logger {
+	return logger.WithFields(logrus.Fields{
+		"JobRun.ID":   j.ID,
+		"JobRun.Name": j.Name.String,
+		"JobRun.Job":  j.Job,
+	})
+}
+
+func (j *JobRun) run(jobsd *JobsD) (int64, error) {
+	if err := j.check(jobsd.jobs, jobsd.schedules); err != nil {
+		return 0, err
+	}
+
+	j.RunAt = time.Now().Add(j.Delay)
+	if j.needsScheduling() {
+		schedule, exists := jobsd.schedules[j.Schedule.String]
+		if !exists {
+			return 0, errors.New("cannot schedule job run, schedule does not exist")
+		}
+		j.RunAt = schedule(j.RunAt)
+	}
+
+	err := j.insertGet(jobsd.db)
+	if err != nil {
+		return 0, err
+	}
+	jobsd.log.WithFields(map[string]interface{}{
```

Comment:
@cleancode4ever this may surprise you but this is intentional. I purposefully decoupled the logging functions so that you can use any logging library.

See https://github.com/simpleframeworks/logc . This library allows me to change out the logging.

Reply:
Great! Never thought of this.